sys/userlib/src/arch/riscv64.rs

```diff
@@ -0,0 +1,397 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! User application architecture support stubs for RISC-V
+//!
+//! See the note on syscall stubs at the top of the userlib module for
+//! rationale.
+
+use crate::*;
+use core::arch::asm;
+
+/// This is the entry point for the kernel. Its job is to set up our memory
+/// before jumping to user-defined `main`.
+#[doc(hidden)]
+#[no_mangle]
+#[link_section = ".text.start"]
+#[naked]
+pub unsafe extern "C" fn _start() -> ! {
+    // Provided by the user program:
+    extern "Rust" {
+        fn main() -> !;
+    }
+
+    asm!("
+        # Copy data initialization image into data section.
+        la t0, __edata       # upper bound in t0
+        la t1, __sidata      # source in t1
+        la t2, __sdata       # dest in t2
+        j 1f
+
+    2:  ld s3, (t1)
+        add t1, t1, 8
+        sd s3, (t2)
+        add t2, t2, 8
+
+    1:  bne t2, t0, 2b
+
+        # Zero BSS
+        la t0, __ebss        # upper bound in t0
+        la t1, __sbss        # base in t1
+        j 1f
+
+    2:  sd zero, (t1)
+        add t1, t1, 8
+
+    1:  bne t1, t0, 2b
+        j {main}
+        ",
+        main = sym main,
+        options(noreturn),
+    )
+}
+
+/// Core implementation of the REFRESH_TASK_ID syscall.
+#[naked]
+pub(crate) unsafe extern "C" fn sys_refresh_task_id_stub(_tid: u32) -> u32 {
+    asm!("
+        # Load the constant syscall number.
+        li a7, {sysnum}
+
+        # To the kernel!
+        ecall
+
+        # Results are placed into the correct registers by the kernel, we can
+        # just return now.
+        ret
+        ",
+        sysnum = const Sysnum::RefreshTaskId as u32,
+        options(noreturn),
+    )
+}
+
+/// Core implementation of the SEND syscall.
+#[naked]
+pub(crate) unsafe extern "C" fn sys_send_stub(
+    _args: &mut SendArgs<'_>,
+) -> RcLen {
+    asm!("
+        # Load in args from the struct (SendArgs).
+        ld a6, 5*8(a0)    # lease_len
+        ld a5, 4*8(a0)    # lease_ptr
+        ld a4, 3*8(a0)    # incoming_len
+        ld a3, 2*8(a0)    # incoming_ptr
+        ld a2, 1*8(a0)    # outgoing_len
+        ld a1, 0*8(a0)    # outgoing_ptr
+        lw a0, 6*8(a0)    # packed_target_operation
+
+        # Load the constant syscall number.
+        li a7, {sysnum}
+
+        # To the kernel!
+        ecall
+
+        slli a1, a1, 0x20
+        or a0, a0, a1
```

Comment:
Why was this added here and at lines 212 and 243?

Reply:
The kernel places `resp` in ret0 and `len` in ret1 when returning back from syscall. Within userlib arch-agnostic code, `sys_send()` expects RcLen as return value from `sys_send_stub()` which is provided by arch-dependent code. RcLen is a struct with a single `u64` type. `sys_send()` takes this RcLen and calls `from()` to convert it into `(u32, usize)`.

So, on 32-bit arch, the `from()` expects RcLen to be held in ret0 and ret1. However, on 64-bit arch, the `from()` expects RcLen to be held in a single ret0. That is why RV64 is applying the transformation to put together the two 32-bit ret values into a single u64.

This is what `from()` looks like on rv64:
```
000000009002311e <from>:
    9002311e:   02055593                srli    a1,a0,0x20
    90023122:   8082                    ret
```

and this is what it looks like on rv32:
```
200150b2 <from>:
200150b2:       8082                    ret
```

Reply:
I think at some point we should switch to passing RcLen as an argument into `sys_send_stub()` and fill that in from arch-dependent code. Then we can just address individual members directly instead of having to do this weird shifting in arch-dependent code. Much difficult to follow.

Reply:
Raised issue [SW-560](http://go/ji/SW-560).